lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -1873,89 +1871,186 @@ void TFDeabstraction::checkAndCanonicalizeAttributes() {
+      // TODO: improve the diagnostic to talk about the parameter label in
+      // the user code, not the internal op attribute.  The bookkeeping for
+      // this isn't obvious though.
+      diagnoseInvalidAttr("requires a constant argument");
+
+      // If we have more specific information about what went wrong, emit
+      // notes.
+      if (it->second.getKind() == SymbolicValue::Unknown)
+        it->second.emitUnknownDiagnosticNotes(opInfo.inst->getLoc());
+      return;
     }
+
+    // Name mangle the attribute classification into the operand list so we can
+    // know the difference between a shape and an array, and a shape array, etc.
+    auto attrName = operandClass.first.str() +
+       SILTensorOpInfo::getOperandClassSuffix(operandClass.second);
+
+    auto constValue = it->second.cloneInto(allocator);
+    auto attrIdentifier = context.getIdentifier(attrName);
+    attributes.push_back({ attrIdentifier, constValue });
+
+    // Verify that the type of this attribute is ok for the OperandClass we
+    // have.
+    switch (operandClass.second) {
+    case SILTensorOpInfo::OperandClass::Input:
+    case SILTensorOpInfo::OperandClass::InputElt:
+    case SILTensorOpInfo::OperandClass::Normal:  // No modifier.
+      break;
+    case SILTensorOpInfo::OperandClass::DType:   // This integer value is a dtype.
+      if (constValue.getKind() != SymbolicValue::Integer)
+        return diagnoseInvalidAttr("requires a constant integer");
+      break;
+    case SILTensorOpInfo::OperandClass::Shape:
+    case SILTensorOpInfo::OperandClass::Array:
+    case SILTensorOpInfo::OperandClass::ShapeArray:
+    case SILTensorOpInfo::OperandClass::ArrayElement:
+      // Match logic from checkAndDiagnoseOperands.
+      assert(0 && "FIXME: array constant exprs aren't handled yet");
+    case SILTensorOpInfo::OperandClass::Tensor:
+      // FIXME: Support array constants.
+      if (constValue.getKind() != SymbolicValue::Integer &&
+          constValue.getKind() != SymbolicValue::Float)
+        return diagnoseInvalidAttr("requires a constant that is an integer,"
+                                   " floating point, or array thereof");
+      break;
+    }
+  }
+
+  // Okay, if we got this far then we have all valid attributes and inputs.
+  // Figure out the result list.
+  SmallVector<SILType, 4> resultTypes;
+  if (auto tuple = inst->getType().getAs<TupleType>()) {
+    for (auto elt : tuple->getElements()) {
+      auto eltTy = elt.getType()->getCanonicalType();
+      resultTypes.push_back(SILType::getPrimitiveObjectType(eltTy));
+    }
+  } else {
+    resultTypes.push_back(inst->getType());
+  }
+
+  SILBuilder B(opInfo.inst);
+  auto op = B.createGraphOperation(inst->getLoc(),
+                                   context.getIdentifier(opName), inputs,
+                                   attributes, resultTypes);
+  op->setDebugLocation(inst->getDebugLocation());
+  
+  if (auto tuple = inst->getType().getAs<TupleType>()) {
```

Comment:
Maybe there's no need to check for tuple type? You could use `op->getResults()` directly and determine whether you have multiple results based on that.

Reply:
Builtins always have a single result.  This code is mapping from that world to the graph_op world that have multiple results.